engine/backends/coq/coq/coq_backend.ml

```diff
@@ -88,598 +104,2219 @@ module Context = struct
+
+      method impl_expr_kind_Parent ~impl:_ ~ident:_ =
+        default_document_for "impl_expr_kind_Parent"
+
+      method impl_expr_kind_Projection ~impl:_ ~item:_ ~ident:_ =
+        default_document_for "impl_expr_kind_Projection"
+
+      method impl_expr_kind_Self = default_document_for "impl_expr_kind_Self"
+      method impl_ident ~goal ~name:_ = goal#p
+      (* string name ^^ space ^^ colon ^^ space ^^ goal#p *)
+      (* TODO: include names and do something about numbered names of instance *)
+
+      method impl_item ~ii_span:_ ~ii_generics:_ ~ii_v ~ii_ident ~ii_attrs:_ =
+        ii_ident#p (* ^^ ii_generics#p *) ^^ space
+        ^^ string ":=" ^^ space ^^ ii_v#p ^^ semi
+
+      method impl_item'_IIFn ~body ~params =
+        if List.length params == 0 then body#p
+        else
+          string "fun" ^^ space
+          ^^ concat_map (fun x -> x#p ^^ space) params
+          ^^ string "=>"
+          ^^ nest 2 (break 1 ^^ body#p)
+
+      method impl_item'_IIType ~typ ~parent_bounds:_ = typ#p
+
+      method item ~v ~span:_ ~ident:_ ~attrs:_ =
+        if is_document_empty v#p then empty else v#p ^^ break 1
+
+      method item'_Alias ~super:_ ~name ~item =
+        string "Notation" ^^ space ^^ string "\"'" ^^ name#p ^^ string "'\""
+        ^^ space ^^ string ":=" ^^ space ^^ parens item#p ^^ dot
+
+      method item'_Fn ~super ~name ~generics ~body ~params ~safety:_ =
+        (* TODO: Why is type not available here ? *)
+        let is_rec =
+          Set.mem
+            (U.Reducers.collect_concrete_idents#visit_expr () body#v)
+            name#v
+        in
+        let typ =
+          self#_do_not_override_lazy_of_ty AstPos_item'_Fn_body body#v.typ
+        in
+
+        let get_expr_of kind f : document =
+          Attrs.associated_expr kind super.attrs
+          |> Option.map ~f:(self#entrypoint_expr >> f)
+          |> Option.value ~default:empty
+        in
+        let requires =
+          get_expr_of Requires (fun x ->
+              x ^^ space ^^ string "=" ^^ space ^^ string "true")
+        in
+        let ensures =
+          get_expr_of Ensures (fun x ->
+              x ^^ space ^^ string "=" ^^ space ^^ string "true")
+        in
+
+        let is_lemma = Attrs.lemma super.attrs in
+        if is_lemma then
+          CoqNotation.lemma name#p generics#p
+            (List.map ~f:(fun x -> x#p) params)
+            (requires ^^ space ^^ !^"->" ^^ break 1 ^^ ensures)
+        else if is_rec then
+          CoqNotation.fixpoint name#p generics#p
+            (List.map ~f:(fun x -> x#p) params
+            @
+            if is_document_empty requires then []
+            else [ string "`" ^^ braces requires ])
+            typ#p body#p (* ^^ TODO: ensures? *)
+        else
+          CoqNotation.definition name#p generics#p
+            (List.map ~f:(fun x -> x#p) params
+            @
+            if is_document_empty requires then []
+            else [ string "`" ^^ braces requires ])
```

Comment:
With requires being a option, that will be easy to replace those `is_document_empty` with a match